text/3396-extern-types-v2.md

```diff
@@ -0,0 +1,249 @@
+
+In the Rust 2021 edition and earlier `T: Sized` implies that `T` is "statically sized" and "statically aligned" and `T: ?Sized` implies that `T` is "metadata sized" and "metadata aligned".
+
+In the Rust 2024 edition and later `T: Sized` means the same but `T: ?Sized` implies that `T` is "unknown sized" and "unknown aligned".
+A `MetaSized` bound can be introduced to regain the previous meaning.
+
+## `MetaSized` trait
+[metasized-trait]: #metasized-trait
+
+This introduces a new trait `core::marker::MetaSized` that represents a type that is "metadata sized" and "metadata aligned":
+```rust
+#[lang = "meta_sized"]
+trait MetaSized {}
+```
+This trait is automatically implemented for all types except extern types.
+
+All types that implement `MetaSized` can be placed as the last field in a struct because the compiler can emit code to determine the offset of the field purely from the pointer metadata.
+At the time of writing all `MetaSized` types are either statically aligned, or are trait objects and have their alignment in their vtable.
+This RFC does not propose changing codegen for computing the offset as it does not introduce any new `MetaSized` types.
+
+All locations in the standard library that use `?Sized` bounds need to be reviewed when migrating to the 2024 edition and replaced with either `?Sized` or `?Sized + MetaSized` as appropriate. Some examples:
+```rust
+pub fn size_of_val<T: ?Sized + MetaSized>(val: &T) -> usize
+```
+This requires `MetaSized` because the size must be known at run time.
+
+```rust
+pub struct Box<T: ?Sized + MetaSized, A: Allocator = Global>(Unique<T>, A);
+```
+This requires `MetaSized` because the `Box` must be able to determine the Layout of the type at run time to allocate and free the backing memory.
+
+```rust
+impl<A: ?Sized, B: ?Sized> const PartialEq<&B> for &A
+where
+    A: ~const PartialEq<B>,
+```
+This does not require `MetaSized` because the references to `A` and `B` always remain behind a pointer and the size and alignment is not required to call a function on the type.
+
+## Extern types
+[extern-types]: #extern-types
+
+Extern types are defined as above, they are thin DSTs, that is their metadata is `()`. They cannot ever exist except behind a pointer, and so attempts to dereference them fail at compile time similar to trait objects.
+
+`repr` attributes are not permitted on extern types as none of the existing representations are applicable.
+
+Extern types do not automatically implement any traits (except `Pointee`), but users can manually implement, for example, `Send`.
+This does mean that all extern types will be `!Freeze` as it is private, however this is a compiler internal that is only used for optimisations, so this only causes some potential missed optimisations.
+
+An extern type can be included in a `repr(transparent)` struct as it is always at offset 0.
+The transparent struct is then also a thin DST and inherits traits as normal.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
```

Comment:
Another drawback is that this RFC rejects a type that is used widely inside rustc:
```rust
extern "C" {
    type OpaqueListContents;
}

pub struct List<T> {
    len: usize,
    data: [T; 0],
    opaque: OpaqueListContents,
}
```
This must definitely be mentioned in the RFC. Ideally there is some kind of proposal for what to do with this type...

Reply:
A simple suggestion: allow structs with extern tails, but don't allow computing the offset of (or taking a reference of) the extern tail. This is sufficient for the rustc impl, which uses the `data` field rather than the `opaque` field.

A less simple suggestion is a way to specify a known alignment to use for an extern type; this could either be via `repr(align = N)` or `unsafe impl marker::Aligned { const ALIGN: usize = N; }`.